src/shogun/structure/GEMPLP.h

```diff
@@ -0,0 +1,160 @@
+		{}
+
+		/** maximum number of outer iterations*/
+		int32_t m_max_iter;
+		/** threshold of the delta objective value */
+		float64_t m_obj_del_thr;
+		/** threshold of objective gap betweeb current and best integer assignment */
+		float64_t m_int_gap_thr;
+	};
+
+public:
+	/** Constructor */
+	CGEMPLP();
+
+	/** Constructor
+	 *
+	 * @param fg factor graph
+	 * @param param parameters
+	 */
+	CGEMPLP(CFactorGraph* fg, Parameter param = Parameter());
+
+	/** Destructor */
+	virtual ~CGEMPLP();
+
+	/** @return class name */
+	virtual const char* get_name() const
+	{
+		return "GEMPLP";
+	}
+
+	/** Inference
+	 *
+	 * @param assignment the assignment
+	 * @return the total energy after doing inference
+	 */
+	virtual float64_t inference(SGVector<int32_t> assignment);
+
+private:
+	/** Initialize GEMPLP with factor graph */
+	void init();
+	
+	/** Message updating on a clique
+	 *
+	 * Please refer to "GEMPLP" in NIPS paper of
+	 * A. Globerson and T. Jaakkola [1] for more details.
+	 *	
+	 */
+	void update_messages(int32_t id_clique);	
+
+public:
+	/** Computer the maximum value along the sub-dimension
+	 *
+	 * @param tar_arr target nd array
+	 * @param subset_inds sub-dimension indices
+	 * @param max_res the result nd array
+	 */
+	void max_in_subdimension(SGNDArray<float64_t> tar_arr, SGVector<int32_t> &subset_inds, SGNDArray<float64_t> &max_res) const;
+
+	/** Find separators between clqiues
+	 *
+	 * @param clique_A clique A
+	 * @param clique_B clique B
+	 * @return index in all separators
+	 */
+	int32_t find_separator_index(SGVector<int32_t> clique_A, SGVector<int32_t> clique_B);
+
+	/** Convert original energies to potentials of the cluster
+	 * GEMPLP objective function is a maximazation function, we use - energy
+	 * as potential. The indices of the table factor energy also need to be
+	 * converted to the order of the nd array.
+	 *
+	 * @param factor factor which contains energy
+	 * @return potential of the cluster in MPLP
+	 */
+	SGNDArray<float64_t> convert_message(CFactor* factor);
+
+	/** Insert the new value into a vector
+	 *
+	 * @param v the target vector
+	 * @param k current value
+	 */
+	void exist_or_pushback(vector<int32_t>& v, int32_t k);
+
+public:
+	/** GEMPLP parameter */
+	Parameter m_param;
+	/** all factors in the graph*/
+	CDynamicObjectArray* m_factors;
+	/** all separators */
+	vector<SGVector<int32_t> > m_all_separators;
+	/** the separator indices (node indices) on each clique */
+	vector<vector<int32_t> > m_clique_separators;
+	/** the indices (orders in the clique) of the separators on each clique */
+	vector<vector<SGVector<int32_t> > > m_clique_inds_separators;
+	/** store the sum of messages into separators */
+	vector<SGNDArray<float64_t> > m_msgs_into_separators;
+	/** store the messages from cluster to separator */
+	vector<vector<SGNDArray<float64_t> > > m_msgs_from_cluster;
+	/** store the original (-) energy of the factors */
+	vector<SGNDArray<float64_t> > m_theta_cluster;
```

Comment:
-> m_region_theta
I know they call m_region_lambda, but lambda was used in the paper differently .